File: algorithms/809/809.cpp
#include <cassert>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  int expressiveWords(string S, vector<string>& words) {
    int count = 0;
    for (const auto& W : words) {
      if (check(S, W)) {
        count++;
      }
    }
    return count;
  }

  bool check(const string& S, const string& W) {
    // W: hello
    // S: heeellooo
    int s_idx = 0;
    int w_idx = 0;
    while (true) {
      // W: helo
      //        x
      // S: heeellooo
      //             x
      if (s_idx == S.size() && w_idx == W.size()) {
        //printf("\n");
        return true;
      }
      // W: helo
      //     x
      // S: heeellooo
      //     x
      if (w_idx < W.size() && W[w_idx] == S[s_idx]) {
        //printf("w_idx: %d, s_idx: %d, w_idx++\n", w_idx, s_idx);
        w_idx++;
      }
      // W: helo
      //      x
      // S: heeellooo
      //      x
      else if (s_idx > 0 && S[s_idx] == S[s_idx+1] && s_idx < S.size() && S[s_idx] == S[s_idx-1]) {
        //printf("w_idx: %d, s_idx: %d, s_idx++\n", w_idx, s_idx);
        s_idx++;
      }
      // W: helo
      //       x
      // S: heeellooo
      //        x
      else if (!(s_idx > 1 && S[s_idx] == S[s_idx-1] && S[s_idx] == S[s_idx-2])) {
        //printf("w_idx: %d, s_idx: %d, break\n", w_idx, s_idx);
        break;
      }
      //printf("w_idx: %d, s_idx: %d, s_idx++\n", w_idx, s_idx);
      s_idx++;
    }
    //printf("\n");
    return false;
  }
};


int main() {
  Solution * sol = new Solution();
  string S = "heeellooo";
  int got;
  vector<string> words;

  words = {"hello"};
  got = sol->expressiveWords(S, words);
  assert(got == 1);

  words = {"hi"};
  got = sol->expressiveWords(S, words);
  assert(got == 0);

  words = {"helo"};
  got = sol->expressiveWords(S, words);
  assert(got == 0);

  words = {"e"};
  got = sol->expressiveWords(S, words);
  assert(got == 0);

  words = {"helllo"};
  got = sol->expressiveWords(S, words);
  assert(got == 0);
  delete sol;
}

File: algorithms/809/809.py
"""
Let S be a given word and W be a word contained in `words` variable.
When the number of characters in a group in S is
    1, the number in W must be 1
    2, the number in W must be 2
    3, the number in W can be 1,2,3
    4, the number in W can be 1,2,3,4
    5, ...
"""

class Solution:
    def breakDownString(self, s):
        """
        :param s: a string (e.x. "heeellooo")
        :type s: str
        :returns: length of each character in a given string
                  (e.x. [('h', 1), ('e', 3), ('l', 2), ('o', 3)])
        :rtype: List[(str, int)]
        """
        prev_c = None
        count = 0
        l = []
        for c in s:
            if prev_c is not None and c != prev_c:
                l.append((prev_c, count))
                count = 0
            count += 1
            prev_c = c
        l.append((prev_c, count))
        return l

    def expressiveWords(self, S, words):
        """
        :type S: str
        :type words: List[str]
        :rtype: int
        """
        l_S = self.breakDownString(S)
        count = 0
        for W in words:
            l_W = self.breakDownString(W)
            ok = True
            if len(l_S) != len(l_W):
                continue
            for pair_S, pair_W in zip(l_S, l_W):
                if pair_S[0] != pair_W[0]:
                    ok = False
                    break
                if pair_S[1] == 1 and pair_W[1] != 1:
                    ok = False
                    break
                if pair_S[1] == 2 and pair_W[1] != 2:
                    ok = False
                    break
                if pair_S[1] >= 3 and pair_W[1] > pair_S[1]:
                    ok = False
                    break
            if ok is True:
                count += 1
        return count


if __name__ == '__main__':
    sol = Solution()
    S = "heeellooo"
    words = ["hello", "hi", "helo", "heello", "heeello"]
    got = sol.expressiveWords(S, words)
    assert got == 3     # hello, heello, heeello
    words = ['heko']
    got = sol.expressiveWords(S, words)
    assert got == 0
